### PR TITLE
UCP/PROTO: Print endpoint's protocols configuration as a table

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Improve visibility to UCX protocols selection and switch-over thresholds

### CODE REVIEW

**Comment by shamisp** [comment]
Can you add output sample pls ? thanks

**Comment by yosefe** [comment]
> Can you add output sample pls ? thanks

<details>
<summary>
[click to expand] $ mpirun -x UCX_PROTO_ENABLE=y -x UCX_LOG_LEVEL=info -np 2 --map-by node osu_bw D D
</summary>
The performance below is limited by PCI p2p bandwidth (~3500MBs), before switching to 3-stage pipeline (~10000 MBs)

```
[1645019895.753272] [vulcan01:22386:0]     ucp_context.c:1779 UCX  INFO  UCP version is 1.13 (release 0)
[1645019895.865222] [vulcan01:22386:0]          parser.c:1918 UCX  INFO  UCX_* env variables: UCX_PROTO_ENABLE=y UCX_LOG_LEVEL=info
[1645019897.339508] [vulcan04:13202:0]     ucp_context.c:1779 UCX  INFO  UCP version is 1.13 (release 0)
[1645019897.432419] [vulcan04:13202:0]          parser.c:1918 UCX  INFO  UCX_* env variables: UCX_PROTO_ENABLE=y UCX_LOG_LEVEL=info
[1645019897.434647] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+---------------------------------------------------------------------------------------------------+
[1645019897.434657] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  | mpi ep_cfg[2] | tagged message by ucp_tag_send*() from host memory                                                |
[1645019897.434661] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------+------------------------------------------------+
[1645019897.434665] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |       0..8184 | eager short                                      | self/memory0                                   |
[1645019897.434669] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |    8185..9806 | eager copy-in copy-out                           | self/memory0                                   |
[1645019897.434673] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |     9807..inf | (?) rendezvous zero-copy flushed write to remote | 55% on knem/memory and 45% on rc_mlx5/mlx5_0:1 |
[1645019897.434676] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------+------------------------------------------------+
[1645019897.434745] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+---------------------------------------------------------------------------------------------------+
[1645019897.434760] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  | mpi ep_cfg[2] | tagged message by ucp_tag_send*() from host memory                                                |
[1645019897.434765] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------+------------------------------------------------+
[1645019897.434779] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |       0..8184 | eager short                                      | self/memory0                                   |
[1645019897.434783] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |    8185..9806 | eager copy-in copy-out                           | self/memory0                                   |
[1645019897.434788] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |     9807..inf | (?) rendezvous zero-copy flushed write to remote | 55% on knem/memory and 45% on rc_mlx5/mlx5_0:1 |
[1645019897.434793] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------+------------------------------------------------+
# MPI_Init() took 1779.033 msec
# OSU MPI-CUDA Bandwidth Test v5.6.2
# Send Buffer on DEVICE (D) and Receive Buffer on DEVICE (D)
# Size    Bandwidth (MB/s)
[1645019897.509095] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------------------+
[1645019897.509105] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  | mpi ep_cfg[3] | tagged message by ucp_tag_send*() from host memory           |
[1645019897.509110] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------------+------------------+
[1645019897.509114] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |       0..2007 | eager short                               | rc_mlx5/mlx5_0:1 |
[1645019897.509119] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |    2008..8246 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645019897.509123] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |   8247..17297 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645019897.509131] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------------+------------------+
[1645019897.516468] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+------------------------------------------------------------------------------------+
[1645019897.516481] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  | mpi ep_cfg[3] | tagged message by ucp_tag_send*() from cuda/GPU0                                   |
[1645019897.516486] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+-----------------------------------------------------------------+------------------+
[1645019897.516490] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |       0..8246 | eager zero-copy copy-out                                        | rc_mlx5/mlx5_0:1 |
[1645019897.516494] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |  8247..811555 | eager zero-copy copy-out                                        | rc_mlx5/mlx5_0:1 |
[1645019897.516498] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |   811556..inf | (?) rendezvous pipeline cuda_copy, fenced write to remote, cuda | rc_mlx5/mlx5_0:1 |
[1645019897.516502] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+-----------------------------------------------------------------+------------------+
[1645019897.519901] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------------------+
[1645019897.519920] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  | mpi ep_cfg[3] | tagged message by ucp_tag_send*() from host memory           |
[1645019897.519924] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------------+------------------+
[1645019897.519927] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |       0..2007 | eager short                               | rc_mlx5/mlx5_0:1 |
[1645019897.519931] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |    2008..8246 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645019897.519934] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |   8247..17297 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645019897.519937] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |    17298..inf | (?) rendezvous zero-copy read from remote | rc_mlx5/mlx5_0:1 |
[1645019897.519941] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------------+------------------+
1                 1.88
2                 3.81
4                 7.57
8                15.44
16               29.96
32               59.88
64              116.66
128             227.98
256             442.42
512             855.68
1024           1566.62
2048           2528.01
4096           3407.60
8192           3411.18
16384          3689.44
32768          3575.94
65536          3576.68
131072         3581.93
262144         3589.99
524288         3558.45
[1645019898.033837] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------------------------------+
[1645019898.033847] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  | mpi ep_cfg[3] | rendezvous data fetch() into cuda/GPU0 from cuda/dev[0]                  |
[1645019898.033851] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------------------------+------------------+
[1645019898.033854] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |         1..64 | (?) zero-copy fenced write to remote                  | rc_mlx5/mlx5_0:1 |
[1645019898.033857] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |    65..563180 | zero-copy read from remote                            | rc_mlx5/mlx5_0:1 |
[1645019898.033860] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  |   563181..inf | pipeline cuda_copy, fenced write to remote, cuda_copy | rc_mlx5/mlx5_0:1 |
[1645019898.033863] [vulcan04:13202:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------------------------+------------------+
[1645019898.067260] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+--------------------------------------------------------+
[1645019898.067270] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  | mpi ep_cfg[3] | rendezvous data send(multi) from cuda/GPU0 to host     |
[1645019898.067275] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------+------------------+
[1645019898.067279] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |      0..45850 | zero-copy fenced write to remote    | rc_mlx5/mlx5_0:1 |
[1645019898.067282] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |   45851..512K | cuda_copy, fenced write to remote   | rc_mlx5/mlx5_0:1 |
[1645019898.067286] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  |   524289..inf | pipeline cuda_copy, write to remote | rc_mlx5/mlx5_0:1 |
[1645019898.067290] [vulcan01:22386:0]     proto_debug.c:406  UCX  INFO  +---------------+-------------------------------------+------------------+
1048576       10129.18
2097152       10115.62
4194304       10137.78
```

</details>

File: src/ucp/proto/proto_select.h [hunk 1]
```diff
@@ -26,6 +26,11 @@
 #define UCP_PROTO_SELECT_OP_FLAG_PPLN_FRAG (UCP_PROTO_SELECT_OP_FLAGS_BASE << 0)
 
 
+/* Select a protocol as part of performance estimation of another protoocol,
```
Comment (hoopoepg) [inline src/ucp/proto/proto_select.h:29]: protOOcol?

**Review by hoopoepg — APPROVED** [review]

**Comment by yosefe** [comment]
> > Can you add output sample pls ? thanks
> 
> [click to expand] $ mpirun -x UCX_PROTO_ENABLE=y -x UCX_LOG_LEVEL=info -np 2 --map-by node osu_bw D D

@shamisp  another option is to print a more compact log line, using a separate env var, like this:

<details>
<summary>
[click me] $ mpirun -x UCX_PROTO_ENABLE=y -x UCX_PROTO_INFO=y -np 2 --map-by node osu_bw D D
</summary>

```
[1645027038.617078] [vulcan04:22884:0]   +---------------+---------------------------------------------------------------------------------------------------+
[1645027038.617101] [vulcan04:22884:0]   | mpi ep_cfg[2] | tagged message by ucp_tag_send*() from host memory                                                |
[1645027038.617104] [vulcan04:22884:0]   +---------------+--------------------------------------------------+------------------------------------------------+
[1645027038.617107] [vulcan04:22884:0]   |       0..8184 | eager short                                      | self/memory0                                   |
[1645027038.617110] [vulcan04:22884:0]   |    8185..9806 | eager copy-in copy-out                           | self/memory0                                   |
[1645027038.617112] [vulcan04:22884:0]   |     9807..inf | (?) rendezvous zero-copy flushed write to remote | 55% on knem/memory and 45% on rc_mlx5/mlx5_0:1 |
[1645027038.617115] [vulcan04:22884:0]   +---------------+--------------------------------------------------+------------------------------------------------+
[1645027038.617307] [vulcan01:31965:0]   +---------------+---------------------------------------------------------------------------------------------------+
[1645027038.617337] [vulcan01:31965:0]   | mpi ep_cfg[2] | tagged message by ucp_tag_send*() from host memory                                                |
[1645027038.617341] [vulcan01:31965:0]   +---------------+--------------------------------------------------+------------------------------------------------+
[1645027038.617344] [vulcan01:31965:0]   |       0..8184 | eager short                                      | self/memory0                                   |
[1645027038.617348] [vulcan01:31965:0]   |    8185..9806 | eager copy-in copy-out                           | self/memory0                                   |
[1645027038.617351] [vulcan01:31965:0]   |     9807..inf | (?) rendezvous zero-copy flushed write to remote | 55% on knem/memory and 45% on rc_mlx5/mlx5_0:1 |
[1645027038.617354] [vulcan01:31965:0]   +---------------+--------------------------------------------------+------------------------------------------------+
# MPI_Init() took 1479.255 msec
# OSU MPI-CUDA Bandwidth Test v5.6.2
# Size    Bandwidth (MB/s)
[1645027038.674035] [vulcan01:31965:0]   +---------------+--------------------------------------------------------------+
[1645027038.674043] [vulcan01:31965:0]   | mpi ep_cfg[3] | tagged message by ucp_tag_send*() from host memory           |
[1645027038.674047] [vulcan01:31965:0]   +---------------+-------------------------------------------+------------------+
[1645027038.674049] [vulcan01:31965:0]   |       0..2007 | eager short                               | rc_mlx5/mlx5_0:1 |
[1645027038.674052] [vulcan01:31965:0]   |    2008..8246 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645027038.674055] [vulcan01:31965:0]   |   8247..17297 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645027038.674058] [vulcan01:31965:0]   |    17298..inf | (?) rendezvous zero-copy read from remote | rc_mlx5/mlx5_0:1 |
[1645027038.674060] [vulcan01:31965:0]   +---------------+-------------------------------------------+------------------+
[1645027038.680982] [vulcan01:31965:0]   +---------------+------------------------------------------------------------------------------------+
[1645027038.680993] [vulcan01:31965:0]   | mpi ep_cfg[3] | tagged message by ucp_tag_send*() from cuda/GPU0                                   |
[1645027038.680996] [vulcan01:31965:0]   +---------------+-----------------------------------------------------------------+------------------+
[1645027038.680999] [vulcan01:31965:0]   |       0..8246 | eager zero-copy copy-out                                        | rc_mlx5/mlx5_0:1 |
[1645027038.681001] [vulcan01:31965:0]   |  8247..811555 | eager zero-copy copy-out                                        | rc_mlx5/mlx5_0:1 |
[1645027038.681004] [vulcan01:31965:0]   |   811556..inf | (?) rendezvous pipeline cuda_copy, fenced write to remote, cuda | rc_mlx5/mlx5_0:1 |
[1645027038.681007] [vulcan01:31965:0]   +---------------+-----------------------------------------------------------------+------------------+
[1645027038.693843] [vulcan04:22884:a]   +---------------+--------------------------------------------------------------+
[1645027038.693856] [vulcan04:22884:a]   | mpi ep_cfg[3] | tagged message by ucp_tag_send*() from host memory           |
[1645027038.693858] [vulcan04:22884:a]   +---------------+-------------------------------------------+------------------+
[1645027038.693861] [vulcan04:22884:a]   |       0..2007 | eager short                               | rc_mlx5/mlx5_0:1 |
[1645027038.693863] [vulcan04:22884:a]   |    2008..8246 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645027038.693865] [vulcan04:22884:a]   |   8247..17297 | eager zero-copy copy-out                  | rc_mlx5/mlx5_0:1 |
[1645027038.693867] [vulcan04:22884:a]   |    17298..inf | (?) rendezvous zero-copy read from remote | rc_mlx5/mlx5_0:1 |
[1645027038.693869] [vulcan04:22884:a]   +---------------+-------------------------------------------+------------------+
1                 1.83
2                 3.75
4                 7.45
8                15.14
16               30.02
32               58.95
64              114.62
128             225.86
256             447.85
512             848.30
1024           1545.36
2048           2786.19
4096           3557.48
8192           3587.98
16384          3325.22
32768          3861.98
65536          3594.98
131072         3737.18
262144         3745.48
524288         3726.40
[1645027039.230240] [vulcan04:22884:0]   +---------------+--------------------------------------------------------------------------+
[1645027039.230249] [vulcan04:22884:0]   | mpi ep_cfg[3] | rendezvous data fetch() into cuda/GPU0 from cuda/dev[0]                  |
[1645027039.230252] [vulcan04:22884:0]   +---------------+-------------------------------------------------------+------------------+
[1645027039.230254] [vulcan04:22884:0]   |         1..64 | (?) zero-copy fenced write to remote                  | rc_mlx5/mlx5_0:1 |
[1645027039.230256] [vulcan04:22884:0]   |    65..563180 | zero-copy read from remote                            | rc_mlx5/mlx5_0:1 |
[1645027039.230259] [vulcan04:22884:0]   |   563181..inf | pipeline cuda_copy, fenced write to remote, cuda_copy | rc_mlx5/mlx5_0:1 |
[1645027039.230261] [vulcan04:22884:0]   +---------------+-------------------------------------------------------+------------------+
[1645027039.263024] [vulcan01:31965:0]   +---------------+--------------------------------------------------------+
[1645027039.263033] [vulcan01:31965:0]   | mpi ep_cfg[3] | rendezvous data send(multi) from cuda/GPU0 to host     |
[1645027039.263035] [vulcan01:31965:0]   +---------------+-------------------------------------+------------------+
[1645027039.263038] [vulcan01:31965:0]   |      0..45850 | zero-copy fenced write to remote    | rc_mlx5/mlx5_0:1 |
[1645027039.263041] [vulcan01:31965:0]   |   45851..512K | cuda_copy, fenced write to remote   | rc_mlx5/mlx5_0:1 |
[1645027039.263043] [vulcan01:31965:0]   |   524289..inf | pipeline cuda_copy, write to remote | rc_mlx5/mlx5_0:1 |
[1645027039.263046] [vulcan01:31965:0]   +---------------+-------------------------------------+------------------+
1048576       10122.28
2097152       10093.07
4194304       10099.53
```

</details>

**Comment by shamisp** [comment]
> > > Can you add output sample pls ? thanks
> > 
> > 
> > [click to expand] $ mpirun -x UCX_PROTO_ENABLE=y -x UCX_LOG_LEVEL=info -np 2 --map-by node osu_bw D D
> 
> @shamisp another option is to print a more compact log line, using a separate env var, like this:
> 
>  [click me] $ mpirun -x UCX_PROTO_ENABLE=y -x UCX_PROTO_INFO=y -np 2 --map-by node osu_bw D D

Nice. Can we document in readthedocs FAQ as an example how to extract this info ?

**Comment by yosefe** [comment]
@brminich @hoopoepg can you pls take a look on the latest version?

**Review by hoopoepg — APPROVED** [review]

**Comment by yosefe** [comment]
@brminich @hoopoepg can you pls check the latest 3 changes?

**Review by hoopoepg — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Pull request merged**